### PR TITLE
Add junit rule that reports if test has no assertion

### PR DESCRIPTION
**Pull request by arunvelsriram**
Related issue: #379 

I have added one of the junit rule (2nd one) mentioned in https://github.com/arturbosch/detekt/issues/379#issuecomment-539077928

Am working on adding other rules as well. Since this is where we are introducing junit rules I have added one rule only to get early feedback from the community.

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        Severity.CodeSmell,
+        "A JUnit test should have at least one assertion",
+        Debt.TEN_MINS
+    )
+
+    override fun visitNamedFunction(function: KtNamedFunction) {
+        super.visitNamedFunction(function)
+        function.annotationEntries.find {
+            it.text == TEST_ANNOTATION
+        }?.let {
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: Please prefer normal if-statements for multiline bodies

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            val body = function.bodyExpression?.text
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: You could introduce an early return here with `?: return`

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            if (body?.contains(ASSERTION_PATTERN, ignoreCase = true) == false) {
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: The assertion pattern should be a configurable regex.
The user may use a library like `strikt` where he would use `expectThat` syntax instead of `assert`.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                        Entity.from(function),
+                        issue.description
+                    )
+                )
+            }
+        }
+    }
+
+    companion object {
+        private const val TEST_ANNOTATION = "@Test"
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: Not sure about if this should be a configurable regex

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.RuleSet
+import io.gitlab.arturbosch.detekt.api.RuleSetProvider
+import io.gitlab.arturbosch.detekt.rules.junit.TestWithoutAssertion
+
+/**
+ * Rules in this rule set report issues related to JUnit tests
+ *
+ */
+class JUnitProvider : RuleSetProvider {
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt]: @schalkms @3flex should we prefer a `TestProvider` and have rule names like `JUnitTestWithoutAssertion`?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertionSpec.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package io.gitlab.arturbosch.detekt.rules.junit
+
+import io.gitlab.arturbosch.detekt.test.lint
+import org.assertj.core.api.Assertions.assertThat
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+
+class TestWithoutAssertionSpec : Spek({
+    describe("junit test assertion") {
```
Comment (arturbosch) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertionSpec.kt]: We should add more tests when the assertion pattern is configurable

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt (hunk 5 above)
Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt]: I was thinking something similar, but I wasn't sure what we do about the case where we want an equivalent rule for Spek as well... would we have two rules, `JUnitTestWithoutAssertion` and `SpekTestWithoutAssertion`, or just one `TestWithoutAssertion` that covers both cases and can be extended to support other test engines?

Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt]: Or, we have a `JUnitProvider` and a `SpekProvider` with possibly duplicated rules between them so it's easy to enable only the rulesets that apply to the project (which would typically use a single test engine)

Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt]: I'm okay with both approaches.

File: detekt-cli/src/main/resources/default-detekt-config.yml [hunk 7]
```diff
@@ -285,6 +285,11 @@ formatting:
     active: true
     autoCorrect: true
 
+junit:
```
Comment (schalkms) [inline detekt-cli/src/main/resources/default-detekt-config.yml]: Maybe we should use a more generic name here for all test related rules.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt (hunk 1 above)
Comment (arunvelsriram) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: Done

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt (hunk 2 above)
Comment (arunvelsriram) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: Done

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt (hunk 3 above)
Comment (arunvelsriram) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: Done

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt (hunk 4 above)
Comment (arunvelsriram) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertion.kt]: In JUnit, the annotation is `@Test` always as far as I know.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertionSpec.kt (hunk 6 above)
Comment (arunvelsriram) [reply detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/junit/TestWithoutAssertionSpec.kt]: Done

**Comment by cortinico** [comment]
Just dropping my opinion here in the loop:

Quoting @schalkms 
> Maybe we should use a more generic name here for all test related rules.

I have the feeling that `junit` is definitely a bad name for this rule (multirule) as it will be easily confused with JUnit Rule (see  https://junit.org/junit4/javadoc/4.12/org/junit/Rule.html), a mechanism provided by the JUnit framework to run setup and cleanup statements before JUnit tests.

So reading the name of this PR:

> Add junit rule that reports if test has no assertion

The first impression I had was that we're adding a JUnit Rule to make sure that our Detekt tests contain an assertions and let our internal test suite fail if not.

Long story short: having `testing` is probably better?

**Comment by arunvelsriram** [comment]
@cortinico agreed. `junit` is confusing for users.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt (hunk 5 above)
Comment (arunvelsriram) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt]: if we are changing the `ruleSetId` to `testing` https://github.com/arturbosch/detekt/pull/2019#issuecomment-546880937 then is it safe to assume that this provider will become generic like `TestingProvider`?.

**Comment by arturbosch** [comment]
I agree with all of you that we should have a generic `testing` rule set.
As @schalkms mentioned without resolving all calls and searching for asserts this rule would only be useful inside the test case itself.

We could:
1. resolve all calls inside the file without using the heavy Kotlin compiler resolution - reasoning would be tests should be self contained with private helpers.
2. Work on computing a Call Graph and let rules use it for such cases
3. Rely on Kotlin compilers type resolution (which we should ofc target sooner than later...)

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt (hunk 5 above)
Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/JUnitProvider.kt]: A third approach - don't have a "testing" rule set at all, but include the rules in existing rulesets. This is the approach PMD, Spotbugs and Error Prone take.

In this case it would go in the "bugs" ruleset since it's likely a bug if a test has no assertion.

**Comment by schalkms** [comment]
@arunvelsriram do you have any updates on this PR? Do you need further assistance?

**Comment by arturbosch** [comment]
Unfortunately I propose to close this PR.

This rule is basically a graph reachability problem.
We need to calculate a call graph.
Mark all nodes calling a `assert` or `expect` as "hasAssertion" and traverse all paths from test methods to find out if they eventually lead to a node with "hasAssertion".
We do not have a graph structure now.

Would a `ExplicitAssertion` rule be useful? Every JUnit method need to call a functions starting with name "assert" or "expect"...

**Comment by schalkms** [comment]
> This rule is basically a graph reachability problem.

Agreed. This is currently out of scope what detekt can do.
I propose to close this and have this as a custom rule if you really need this for your code base.

> Would a ExplicitAssertion rule be useful? Every JUnit method need to call a functions starting with name "assert" or "expect"...

I don't think so. There are some many valid cases where you don't need that. For instance, what happens if you only verify the behavior of a unit (e.g. ´verify´).
